INTERNALERROR when error raised in __repr__ with showlocals and verbose output

The small test to reproduce

```python
class Thing:
    def __repr__(self):
        raise NotImplementedError

def test_the_thing():
    t = Thing()
    assert t == 3

def test_something_works():
    assert 1 == 1
```

When running tests with `$ python3 -m pytest` everything works as expected

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py F.                                                          [100%]

======================================== FAILURES ========================================
_____________________________________ test_the_thing _____________________________________

    def test_the_thing():
        t = Thing()
>       assert t == 3
E       assert <[NotImplementedError("") raised in repr()] Thing object at 0x7fd26100aa58> == 3

repr_error_test.py:8: AssertionError
=========================== 1 failed, 1 passed in 0.03 seconds ===========================
```

Adding `-v` and `-l`  -  `$ python3 -m pytest` also works as expected

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py::test_the_thing FAILED                                      [ 50%]
repr_error_test.py::test_something_works PASSED                                [100%]

======================================== FAILURES ========================================
_____________________________________ test_the_thing _____________________________________

    def test_the_thing():
        t = Thing()
>       assert t == 3
E       assert <[NotImplemen...x7fcd312745c0> == 3
E         (pytest_assertion plugin: representation of details failed.  Probably an object has a faulty __repr__.)
E         /home/kondratyev-nv/Development/repr_error/repr_error_test.py:3: NotImplementedError

t          = <[NotImplementedError("") raised in repr()] Thing object at 0x7fcd312745c0>

repr_error_test.py:8: AssertionError
=========================== 1 failed, 1 passed in 0.03 seconds ===========================
```

But increasing verbosity (`$ python3 -m pytest -l -vv`) leads to an INTERNALERROR.

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py::test_the_thing 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 210, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 250, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 271, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 78, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 93, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 175, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 203, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/skipping.py", line 127, in pytest_runtest_makereport
INTERNALERROR>     rep = outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 268, in pytest_runtest_makereport
INTERNALERROR>     longrepr = item.repr_failure(excinfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/python.py", line 837, in repr_failure
INTERNALERROR>     return self._repr_failure_py(excinfo, style=style)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/nodes.py", line 285, in _repr_failure_py
INTERNALERROR>     truncate_locals=truncate_locals,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 556, in getrepr
INTERNALERROR>     return fmt.repr_excinfo(self)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 806, in repr_excinfo
INTERNALERROR>     reprtraceback = self.repr_traceback(excinfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 751, in repr_traceback
INTERNALERROR>     reprentry = self.repr_traceback_entry(entry, einfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 721, in repr_traceback_entry
INTERNALERROR>     localsrepr = self.repr_locals(entry.locals)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 683, in repr_locals
INTERNALERROR>     str_repr = pprint.pformat(value)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 58, in pformat
INTERNALERROR>     compact=compact).pformat(object)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 144, in pformat
INTERNALERROR>     self._format(object, sio, 0, 0, {}, 0)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 161, in _format
INTERNALERROR>     rep = self._repr(object, context, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 393, in _repr
INTERNALERROR>     self._depth, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 405, in format
INTERNALERROR>     return _safe_repr(object, context, maxlevels, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 555, in _safe_repr
INTERNALERROR>     rep = repr(object)
INTERNALERROR>   File "/home/kondratyev-nv/Development/repr_error/repr_error_test.py", line 3, in __repr__
INTERNALERROR>     raise NotImplementedError
INTERNALERROR> NotImplementedError

============================== no tests ran in 0.03 seconds ==============================
```

Also, notice that in the last example no tests are completed. I'm using Ubuntu 18.04, python 3.6.7, and pytest 4.3.0. 

Can you please clarify if this is expected behaviour or a bug?

--- a/AUTHORS
+++ b/AUTHORS
@@ -180,6 +180,7 @@ Nicholas Devenish
 Nicholas Murphy
 Niclas Olofsson
 Nicolas Delaby
+Nikolay Kondratyev
 Oleg Pidsadnyi
 Oleg Sushchenko
 Oliver Bestwalter

--- a/changelog/5089.bugfix.rst
+++ b/changelog/5089.bugfix.rst
@@ -1,0 +1,1 @@
+Fix crash caused by error in ``__repr__`` function with both ``showlocals`` and verbose output enabled.

--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -3,7 +3,6 @@ from __future__ import division
 from __future__ import print_function
 
 import inspect
-import pprint
 import re
 import sys
 import traceback
@@ -18,6 +17,7 @@ import six
 from six import text_type
 
 import _pytest
+from _pytest._io.saferepr import safeformat
 from _pytest._io.saferepr import saferepr
 from _pytest.compat import _PY2
 from _pytest.compat import _PY3
@@ -614,14 +614,11 @@ class FormattedExcinfo(object):
             source = source.deindent()
         return source
 
-    def _saferepr(self, obj):
-        return saferepr(obj)
-
     def repr_args(self, entry):
         if self.funcargs:
             args = []
             for argname, argvalue in entry.frame.getargs(var=True):
-                args.append((argname, self._saferepr(argvalue)))
+                args.append((argname, saferepr(argvalue)))
             return ReprFuncArgs(args)
 
     def get_source(self, source, line_index=-1, excinfo=None, short=False):
@@ -674,9 +671,9 @@ class FormattedExcinfo(object):
                     # _repr() function, which is only reprlib.Repr in
                     # disguise, so is very configurable.
                     if self.truncate_locals:
-                        str_repr = self._saferepr(value)
+                        str_repr = saferepr(value)
                     else:
-                        str_repr = pprint.pformat(value)
+                        str_repr = safeformat(value)
                     # if len(str_repr) < 70 or not isinstance(value,
                     #                            (list, tuple, dict)):
                     lines.append("%-10s = %s" % (name, str_repr))

--- a/src/_pytest/_io/saferepr.py
+++ b/src/_pytest/_io/saferepr.py
@@ -1,6 +1,24 @@
-import sys
+import pprint
 
 from six.moves import reprlib
+
+
+def _call_and_format_exception(call, x, *args):
+    try:
+        # Try the vanilla repr and make sure that the result is a string
+        return call(x, *args)
+    except Exception as exc:
+        exc_name = type(exc).__name__
+        try:
+            exc_info = str(exc)
+        except Exception:
+            exc_info = "unknown"
+        return '<[%s("%s") raised in repr()] %s object at 0x%x>' % (
+            exc_name,
+            exc_info,
+            x.__class__.__name__,
+            id(x),
+        )
 
 
 class SafeRepr(reprlib.Repr):
@@ -33,28 +51,20 @@ class SafeRepr(reprlib.Repr):
         return self._callhelper(repr, x)
 
     def _callhelper(self, call, x, *args):
-        try:
-            # Try the vanilla repr and make sure that the result is a string
-            s = call(x, *args)
-        except Exception:
-            cls, e, tb = sys.exc_info()
-            exc_name = getattr(cls, "__name__", "unknown")
-            try:
-                exc_info = str(e)
-            except Exception:
-                exc_info = "unknown"
-            return '<[%s("%s") raised in repr()] %s object at 0x%x>' % (
-                exc_name,
-                exc_info,
-                x.__class__.__name__,
-                id(x),
-            )
-        else:
-            if len(s) > self.maxsize:
-                i = max(0, (self.maxsize - 3) // 2)
-                j = max(0, self.maxsize - 3 - i)
-                s = s[:i] + "..." + s[len(s) - j :]
-            return s
+        s = _call_and_format_exception(call, x, *args)
+        if len(s) > self.maxsize:
+            i = max(0, (self.maxsize - 3) // 2)
+            j = max(0, self.maxsize - 3 - i)
+            s = s[:i] + "..." + s[len(s) - j :]
+        return s
+
+
+def safeformat(obj):
+    """return a pretty printed string for the given object.
+    Failing __repr__ functions of user instances will be represented
+    with a short exception info.
+    """
+    return _call_and_format_exception(pprint.pformat, obj)
 
 
 def saferepr(obj, maxsize=240):

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -598,6 +598,35 @@ raise ValueError()
         assert reprlocals.lines[2] == "y          = 5"
         assert reprlocals.lines[3] == "z          = 7"
 
+    def test_repr_local_with_error(self):
+        class ObjWithErrorInRepr:
+            def __repr__(self):
+                raise NotImplementedError
+
+        p = FormattedExcinfo(showlocals=True, truncate_locals=False)
+        loc = {"x": ObjWithErrorInRepr(), "__builtins__": {}}
+        reprlocals = p.repr_locals(loc)
+        assert reprlocals.lines
+        assert reprlocals.lines[0] == "__builtins__ = <builtins>"
+        assert '[NotImplementedError("") raised in repr()]' in reprlocals.lines[1]
+
+    def test_repr_local_with_exception_in_class_property(self):
+        class ExceptionWithBrokenClass(Exception):
+            @property
+            def __class__(self):
+                raise TypeError("boom!")
+
+        class ObjWithErrorInRepr:
+            def __repr__(self):
+                raise ExceptionWithBrokenClass()
+
+        p = FormattedExcinfo(showlocals=True, truncate_locals=False)
+        loc = {"x": ObjWithErrorInRepr(), "__builtins__": {}}
+        reprlocals = p.repr_locals(loc)
+        assert reprlocals.lines
+        assert reprlocals.lines[0] == "__builtins__ = <builtins>"
+        assert '[ExceptionWithBrokenClass("") raised in repr()]' in reprlocals.lines[1]
+
     def test_repr_local_truncated(self):
         loc = {"l": [i for i in range(10)]}
         p = FormattedExcinfo(showlocals=True)

--- a/testing/test_session.py
+++ b/testing/test_session.py
@@ -134,6 +134,30 @@ class SessionTests(object):
             != -1
         )
 
+    def test_broken_repr_with_showlocals_verbose(self, testdir):
+        p = testdir.makepyfile(
+            """
+            class ObjWithErrorInRepr:
+                def __repr__(self):
+                    raise NotImplementedError
+
+            def test_repr_error():
+                x = ObjWithErrorInRepr()
+                assert x == "value"
+        """
+        )
+        reprec = testdir.inline_run("--showlocals", "-vv", p)
+        passed, skipped, failed = reprec.listoutcomes()
+        assert (len(passed), len(skipped), len(failed)) == (0, 0, 1)
+        entries = failed[0].longrepr.reprtraceback.reprentries
+        assert len(entries) == 1
+        repr_locals = entries[0].reprlocals
+        assert repr_locals.lines
+        assert len(repr_locals.lines) == 1
+        assert repr_locals.lines[0].startswith(
+            'x          = <[NotImplementedError("") raised in repr()] ObjWithErrorInRepr'
+        )
+
     def test_skip_file_by_conftest(self, testdir):
         testdir.makepyfile(
             conftest="""

